### DISCUSSION TITLE
Generating a secret from secrets-store-csi-driver for Provider

### DISCUSSION THREAD

**Discussion by NeuronButter** [fluxcd/flux2#5012] [Q&A] [answered]
Hi there,
I'm trying to think of a way for Flux to be able to get a secret out of OpenBao (/Vault) to be able to use for the [address](https://fluxcd.io/flux/components/notification/providers/#address) of a Provider. The closest solution to this I've got is using the secrets-store-csi-driver to [sync a secret](https://secrets-store-csi-driver.sigs.k8s.io/topics/sync-as-kubernetes-secret) to be referenced by a [secretRef](https://fluxcd.io/flux/components/notification/providers/#secret-reference), but the only limitation is that it requires the volume to be mounted on a Pod. My thoughts are that it would be mounted on the notification-controller, but I'm not sure on how to extend the controller to be able to consume that volume (which will in turn trigger the secret to generate).
Has anyone encountered such a situation before (not necessarily using the CSI driver either, but just some way of creating a secret for Flux to consume like this)? I'm trying to avoid SOPS and solely use OpenBao for everything, although this is making me consider using SOPS for just this one secret.
Thanks!

Edit: Also looking into just using Vault instead of OpenBao since they still maintain their secrets-operator, was just hoping for an OpenBao solution

**Comment by NeuronButter** [comment, marked as answer]
Couldn't think of a workaround so just using Vault's secret operator with Vault. Would be cool to see Flux support this kind of thing in the future for other CSI drivers that work similarly too though :)

**Reply by stefanprodan** [reply]
A CSI driver works in the context of a pod, it would be so wasteful for Flux to create pods in the cluster just to make some secrets available. All Flux custom resources can refer to existing secrets in etcd, you can either use SOPS and keep the master key in Vault, or you can use SOPS directly with Vault, or you can use external-secrets controllers to sync secrets from Vault into etcd.

**Reply by NeuronButter** [reply]
Ah didn't realise the external-secrets controller could sync Vault -> etcd. I was more thinking about attaching CSI to the notification-controller Pod but suspected that manual changes to Flux objects wouldn't play well in the long run.

**Reply by balonik** [reply]
@stefanprodan That is true, there are alternatives, but why would I have to run another secret management solution if my environment is already using Secret Storage CSI? Secret Storage CSI is getting more and more popular and number of SS CSI agents is increasing.
SOPS and `external-secrets` are a valid alternatives, but it shouldn't be so hard for Flux to use secrets injected into controller pods as files by SS CSI. The password for "initial" repository can be still kept in etcd, but secrets to other gitrepositories, providers, helm repositories which are part of application deployments can be managed by SS CSI.

**Reply by matheuscscp** [reply]
If you mount secrets for all your tenants in the controller, how do you prevent a tenant from referencing a secret that belongs to another tenant? In the `.secretRef` API you have namespace isolation.

**Reply by balonik** [reply]
I understand your concern, however it is valid only in multi-tenant environments. It is not a concern for single tenant clusters. Since cross-namespace referencing is not disabled by default, there is not much of a difference as anyone can reference your objects anyway.
I agree that this can be disabled if cross-namespace is enabled, or simply enabled by a flag.

EDIT: Not to mention how the `external-secrets-operator` works, where people usually have single instance of it with access to any secrets within external store and whoever can reference any secret if he knows the path in external store.
